lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,1039 @@
+	// with the expected request ID from the remote signer is of an
+	// unexpected type.
+	ErrUnexpectedResponse = errors.New("unexpected response type")
+)
+
+// SignCoordinator is an implementation of the signrpc.SignerClient and the
+// walletrpc.WalletKitClient interfaces that passes on all requests to a remote
+// signer. It is used by the watch-only wallet to delegate any signing or ECDH
+// operations to a remote node over a
+// walletrpc.WalletKit_SignCoordinatorStreamsServer stream. The stream is set up
+// by the remote signer when it connects to the watch-only wallet, which should
+// execute the Run method.
+type SignCoordinator struct {
+	// stream is a bi-directional stream between us and the remote signer.
+	stream walletrpc.WalletKit_SignCoordinatorStreamsServer
+
+	// responses is a map of request IDs to response channels. This map
+	// should be populated with a response channel for each request that has
+	// been sent to the remote signer. The response channel should be
+	// inserted into the map before the request is sent.
+	// Any response received over the stream that does not have an
+	// associated response channel in this map is ignored.
+	// The response channel should be removed from the map when the response
+	// has been received and processed.
+	responses map[uint64]chan *walletrpc.SignCoordinatorResponse
+
+	// receiveErrChan is used to signal that the stream with the remote
+	// signer has errored, and we can no longer process responses.
+	receiveErrChan chan error
+
+	// doneReceiving is closed when either party terminates and signals to
+	// any pending requests that we'll no longer process the response for
+	// that request.
+	doneReceiving chan struct{}
+
+	// quit is closed when lnd is shutting down.
+	quit chan struct{}
+
+	// clientConnected is sent over when the remote signer connects.
+	clientConnected chan struct{}
+
+	// nextRequestID keeps track of the next request ID that should
+	// be used when sending a request to the remote signer.
+	nextRequestID uint64
+
+	// requestTimeout is the maximum time we will wait for a response from
+	// the remote signer.
+	requestTimeout time.Duration
+
+	// connectionTimeout is the maximum time we will wait for the remote
+	// signer to connect.
+	connectionTimeout time.Duration
+
+	mu sync.Mutex
+
+	wg sync.WaitGroup
+}
+
+// A compile time assertion to ensure SignCoordinator meets the
+// RemoteSignerRequests interface.
+var _ RemoteSignerRequests = (*SignCoordinator)(nil)
+
+// NewSignCoordinator creates a new instance of the SignCoordinator.
+func NewSignCoordinator(requestTimeout time.Duration,
+	connectionTimeout time.Duration) *SignCoordinator {
+
+	respsMap := make(map[uint64]chan *walletrpc.SignCoordinatorResponse)
+
+	// requestID 1 is reserved for the initial handshake by the remote
+	// signer.
+	nextRequestID := handshakeRequestID + 1
+
+	return &SignCoordinator{
+		responses:         respsMap,
+		receiveErrChan:    make(chan error),
```

Comment:
Does it make sense to buffer the error channels, to make them potentially non-blocking?

Reply:
I don't think that makes any actual difference with the current code actually, but I added it just for extra safety anyways :)